File: ido_sdk_flutter/ido_sdk/protocol_sdk/protocol_channel/ios/Classes/Model/IDOStressSwitchParamModel.swift
//
//  IDOStressSwitchParamModel.swift
//  protocol_channel
//
//  Created by hc on 2024/1/24.
//

import Foundation

// MARK: - IDOStressSwitchParamModel
public class IDOStressSwitchParamModel: NSObject, IDOBaseModel {
    /// Overall switch 1: On 0: Off
    public var onOff: Int
    /// Start time - hour
    public var startHour: Int
    /// Start time - minute
    public var startMinute: Int
    /// End time - hour
    public var endHour: Int
    /// End time - minute
    public var endMinute: Int
    /// Stress reminder switch 1: On 0: Off
    /// Doesn't work if on_off is off
    public var remindOnOff: Int
    /// Reminder interval in minutes, default is 60 minutes
    public var interval: Int
    /// High pressure threshold
    public var highThreshold: Int
    /// Pressure calibration threshold, default is 80
    /// Requires firmware support for setSendCalibrationThreshold
    public var stressThreshold: Int
    /// Notification type
    /// 0: Invalid
    /// 1: Allow notification
    /// 2: Silent notification
    /// 3: Disable notification
    /// Requires firmware support for getPressureNotifyFlagMode
    public var notifyFlag: Int
    
    /// Repeat
    public var repeats: Set<IDOWeek>

    enum CodingKeys: String, CodingKey {
        case onOff = "on_off"
        case startHour = "start_hour"
        case startMinute = "start_minute"
        case endHour = "end_hour"
        case endMinute = "end_minute"
        case remindOnOff = "remind_on_off"
        case interval = "interval"
        case highThreshold = "high_threshold"
        case stressThreshold = "stress_threshold"
        case notifyFlag = "notify_flag"
        case repeats = "repeat"
    }

    public init(onOff: Int, startHour: Int, startMinute: Int, endHour: Int, endMinute: Int, remindOnOff: Int, interval: Int, highThreshold: Int, stressThreshold: Int = 0, notifyFlag: Int, repeats: Set<IDOWeek> = [.monday]) {
        self.onOff = onOff
        self.startHour = startHour
        self.startMinute = startMinute
        self.endHour = endHour
        self.endMinute = endMinute
        self.remindOnOff = remindOnOff
        self.interval = interval
        self.highThreshold = highThreshold
        self.stressThreshold = stressThreshold
        self.notifyFlag = notifyFlag
        self.repeats = repeats
    }
    
    required public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        onOff = try container.decode(Int.self, forKey: .onOff)
        startHour = try container.decode(Int.self, forKey: .startHour)
        startMinute = try container.decode(Int.self, forKey: .startMinute)
        endHour = try container.decode(Int.self, forKey: .endHour)
        endMinute = try container.decode(Int.self, forKey: .endMinute)
        notifyFlag = try container.decode(Int.self, forKey: .notifyFlag)
        remindOnOff = try container.decode(Int.self, forKey: .remindOnOff)
        interval = try container.decode(Int.self, forKey: .interval)
        highThreshold = try container.decode(Int.self, forKey: .highThreshold)
        stressThreshold = try container.decode(Int.self, forKey: .stressThreshold)

        if let value = try container.decodeIfPresent(Int.self, forKey: .repeats) {
            var items = Set<IDOWeek>()
            for i in 0...6 {
                if value & (1 << (i + 1)) != 0 {
                    items.insert(IDOWeek(rawValue: i)!)
                }
            }
            repeats = items
        } else {
            repeats = Set<IDOWeek>()
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(onOff, forKey: .onOff)
        try container.encode(startHour, forKey: .startHour)
        try container.encode(startMinute, forKey: .startMinute)
        try container.encode(endHour, forKey: .endHour)
        try container.encode(endMinute, forKey: .endMinute)
        try container.encode(notifyFlag, forKey: .notifyFlag)
        try container.encode(remindOnOff, forKey: .remindOnOff)
        try container.encode(interval, forKey: .interval)
        try container.encode(highThreshold, forKey: .highThreshold)
        try container.encode(stressThreshold, forKey: .stressThreshold)

        var value = 0
        for item in repeats {
            value |= (1 << (item.rawValue + 1))
        }
        try container.encode(value, forKey: .repeats)
    }

    public func toJsonString() -> String? {
        if let jsonData = try? JSONEncoder().encode(self) {
            let jsonString = String(data: jsonData, encoding: .utf8)
            return jsonString
        }
        return nil
    }
}



// MARK: - IDOStressSwitchParamModelObjc
@objcMembers
public class IDOStressSwitchParamModelObjc: NSObject, Codable {
    /// Overall switch 1: On 0: Off
    public var onOff: Int
    /// Start time - hour
    public var startHour: Int
    /// Start time - minute
    public var startMinute: Int
    /// End time - hour
    public var endHour: Int
    /// End time - minute
    public var endMinute: Int
    /// Stress reminder switch 1: On 0: Off
    /// Doesn't work if on_off is off
    public var remindOnOff: Int
    /// Reminder interval in minutes, default is 60 minutes
    public var interval: Int
    /// High pressure threshold
    public var highThreshold: Int
    /// Pressure calibration threshold, default is 80
    /// Requires firmware support for setSendCalibrationThreshold
    public var stressThreshold: Int
    /// Notification type
    /// 0: Invalid
    /// 1: Allow notification
    /// 2: Silent notification
    /// 3: Disable notification
    /// Requires firmware support for getPressureNotifyFlagMode
    public var notifyFlag: Int
    
    /// Repeat
    public var repeats: [IDOWeekObjc]

    enum CodingKeys: String, CodingKey {
        case onOff = "on_off"
        case startHour = "start_hour"
        case startMinute = "start_minute"
        case endHour = "end_hour"
        case endMinute = "end_minute"
        case remindOnOff = "remind_on_off"
        case interval = "interval"
        case highThreshold = "high_threshold"
        case stressThreshold = "stress_threshold"
        case notifyFlag = "notify_flag"
        case repeats = "repeat"
    }

    public init(onOff: Int, startHour: Int, startMinute: Int, endHour: Int, endMinute: Int, remindOnOff: Int, interval: Int, highThreshold: Int, stressThreshold: Int = 0, notifyFlag: Int, repeats: [IDOWeekObjc]) {
        self.onOff = onOff
        self.startHour = startHour
        self.startMinute = startMinute
        self.endHour = endHour
        self.endMinute = endMinute
        self.remindOnOff = remindOnOff
        self.interval = interval
        self.highThreshold = highThreshold
        self.stressThreshold = stressThreshold
        self.notifyFlag = notifyFlag
        self.repeats = repeats
    }
    
    required public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        onOff = try container.decode(Int.self, forKey: .onOff)
        startHour = try container.decode(Int.self, forKey: .startHour)
        startMinute = try container.decode(Int.self, forKey: .startMinute)
        endHour = try container.decode(Int.self, forKey: .endHour)
        endMinute = try container.decode(Int.self, forKey: .endMinute)
        notifyFlag = try container.decode(Int.self, forKey: .notifyFlag)
        remindOnOff = try container.decode(Int.self, forKey: .remindOnOff)
        interval = try container.decode(Int.self, forKey: .interval)
        highThreshold = try container.decode(Int.self, forKey: .highThreshold)
        stressThreshold = try container.decode(Int.self, forKey: .stressThreshold)

        if let value = try container.decodeIfPresent(Int.self, forKey: .repeats) {
            var items = [IDOWeekObjc]()
            for i in 0...6 {
                if ((value >> i) & 1) > 0 {
                    items.append(IDOWeekObjc(week: IDOWeek(rawValue: i)!))
                }
            }
            repeats = items
        } else {
            repeats = [IDOWeekObjc]()
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(onOff, forKey: .onOff)
        try container.encode(startHour, forKey: .startHour)
        try container.encode(startMinute, forKey: .startMinute)
        try container.encode(endHour, forKey: .endHour)
        try container.encode(endMinute, forKey: .endMinute)
        try container.encode(notifyFlag, forKey: .notifyFlag)
        try container.encode(remindOnOff, forKey: .remindOnOff)
        try container.encode(interval, forKey: .interval)
        try container.encode(highThreshold, forKey: .highThreshold)
        try container.encode(stressThreshold, forKey: .stressThreshold)

        var value = 0
        for item in repeats {
            value |= (1 << (item.week.rawValue + 1))
        }
        try container.encode(value, forKey: .repeats)
    }

    public func toJsonString() -> String? {
        if let jsonData = try? JSONEncoder().encode(self) {
            let jsonString = String(data: jsonData, encoding: .utf8)
            return jsonString
        }
        return nil
    }
}


// MARK: - IDOStressSwitchModel
public class IDOStressSwitchModel: NSObject, IDOBaseModel {
    /// Overall switch 1: On 0: Off
    public var onOff: Int
    /// Start time - hour
    public var startHour: Int
    /// Start time - minute
    public var startMinute: Int
    /// End time - hour
    public var endHour: Int
    /// End time - minute
    public var endMinute: Int
    /// Stress reminder switch 1: On 0: Off
    /// Doesn't work if on_off is off
    public var remindOnOff: Int
    /// Reminder interval in minutes, default is 60 minutes
    public var interval: Int
    /// High pressure threshold
    public var highThreshold: Int
    /// Pressure calibration threshold, default is 80
    /// Requires firmware support for setSendCalibrationThreshold
    public var stressThreshold: Int
    /// Notification type
    /// 0: Invalid
    /// 1: Allow notification
    /// 2: Silent notification
    /// 3: Disable notification
    /// Requires firmware support for getPressureNotifyFlagMode
    public var notifyFlag: Int
    
    /// Repeat
    public var repeats: Set<IDOWeek>

    enum CodingKeys: String, CodingKey {
        case onOff = "on_off"
        case startHour = "start_hour"
        case startMinute = "start_minute"
        case endHour = "end_hour"
        case endMinute = "end_minute"
        case remindOnOff = "remind_on_off"
        case interval = "interval"
        case highThreshold = "high_threshold"
        case stressThreshold = "stress_threshold"
        case notifyFlag = "notify_flag"
        case repeats = "repeat"
    }

    public init(onOff: Int, startHour: Int, startMinute: Int, endHour: Int, endMinute: Int, remindOnOff: Int, interval: Int, highThreshold: Int, stressThreshold: Int = 0, notifyFlag: Int, repeats: Set<IDOWeek> = [.monday]) {
        self.onOff = onOff
        self.startHour = startHour
        self.startMinute = startMinute
        self.endHour = endHour
        self.endMinute = endMinute
        self.remindOnOff = remindOnOff
        self.interval = interval
        self.highThreshold = highThreshold
        self.stressThreshold = stressThreshold
        self.notifyFlag = notifyFlag
        self.repeats = repeats
    }
    
    required public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        onOff = try container.decode(Int.self, forKey: .onOff)
        startHour = try container.decode(Int.self, forKey: .startHour)
        startMinute = try container.decode(Int.self, forKey: .startMinute)
        endHour = try container.decode(Int.self, forKey: .endHour)
        endMinute = try container.decode(Int.self, forKey: .endMinute)
        notifyFlag = try container.decode(Int.self, forKey: .notifyFlag)
        remindOnOff = try container.decode(Int.self, forKey: .remindOnOff)
        interval = try container.decode(Int.self, forKey: .interval)
        highThreshold = try container.decode(Int.self, forKey: .highThreshold)
        stressThreshold = try container.decode(Int.self, forKey: .stressThreshold)

        if let value = try container.decodeIfPresent(Int.self, forKey: .repeats) {
            var items = Set<IDOWeek>()
            for i in 0...6 {
                if value & (1 << (i + 1)) != 0 {
                    items.insert(IDOWeek(rawValue: i)!)
                }
            }
            repeats = items
        } else {
            repeats = Set<IDOWeek>()
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(onOff, forKey: .onOff)
        try container.encode(startHour, forKey: .startHour)
        try container.encode(startMinute, forKey: .startMinute)
        try container.encode(endHour, forKey: .endHour)
        try container.encode(endMinute, forKey: .endMinute)
        try container.encode(notifyFlag, forKey: .notifyFlag)
        try container.encode(remindOnOff, forKey: .remindOnOff)
        try container.encode(interval, forKey: .interval)
        try container.encode(highThreshold, forKey: .highThreshold)
        try container.encode(stressThreshold, forKey: .stressThreshold)

        var value = 0
        for item in repeats {
            value |= (1 << (item.rawValue + 1))
        }
        try container.encode(value, forKey: .repeats)
    }

    public func toJsonString() -> String? {
        if let jsonData = try? JSONEncoder().encode(self) {
            let jsonString = String(data: jsonData, encoding: .utf8)
            return jsonString
        }
        return nil
    }
}



// MARK: - IDOStressSwitchModelObjc
@objcMembers
public class IDOStressSwitchModelObjc: NSObject, Codable {
    /// Overall switch 1: On 0: Off
    public var onOff: Int
    /// Start time - hour
    public var startHour: Int
    /// Start time - minute
    public var startMinute: Int
    /// End time - hour
    public var endHour: Int
    /// End time - minute
    public var endMinute: Int
    /// Stress reminder switch 1: On 0: Off
    /// Doesn't work if on_off is off
    public var remindOnOff: Int
    /// Reminder interval in minutes, default is 60 minutes
    public var interval: Int
    /// High pressure threshold
    public var highThreshold: Int
    /// Pressure calibration threshold, default is 80
    /// Requires firmware support for setSendCalibrationThreshold
    public var stressThreshold: Int
    /// Notification type
    /// 0: Invalid
    /// 1: Allow notification
    /// 2: Silent notification
    /// 3: Disable notification
    /// Requires firmware support for getPressureNotifyFlagMode
    public var notifyFlag: Int
    
    /// Repeat
    public var repeats: [IDOWeekObjc]

    enum CodingKeys: String, CodingKey {
        case onOff = "on_off"
        case startHour = "start_hour"
        case startMinute = "start_minute"
        case endHour = "end_hour"
        case endMinute = "end_minute"
        case remindOnOff = "remind_on_off"
        case interval = "interval"
        case highThreshold = "high_threshold"
        case stressThreshold = "stress_threshold"
        case notifyFlag = "notify_flag"
        case repeats = "repeat"
    }

    public init(onOff: Int, startHour: Int, startMinute: Int, endHour: Int, endMinute: Int, remindOnOff: Int, interval: Int, highThreshold: Int, stressThreshold: Int = 0, notifyFlag: Int, repeats: [IDOWeekObjc]) {
        self.onOff = onOff
        self.startHour = startHour
        self.startMinute = startMinute
        self.endHour = endHour
        self.endMinute = endMinute
        self.remindOnOff = remindOnOff
        self.interval = interval
        self.highThreshold = highThreshold
        self.stressThreshold = stressThreshold
        self.notifyFlag = notifyFlag
        self.repeats = repeats
    }
    
    required public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        onOff = try container.decode(Int.self, forKey: .onOff)
        startHour = try container.decode(Int.self, forKey: .startHour)
        startMinute = try container.decode(Int.self, forKey: .startMinute)
        endHour = try container.decode(Int.self, forKey: .endHour)
        endMinute = try container.decode(Int.self, forKey: .endMinute)
        notifyFlag = try container.decode(Int.self, forKey: .notifyFlag)
        remindOnOff = try container.decode(Int.self, forKey: .remindOnOff)
        interval = try container.decode(Int.self, forKey: .interval)
        highThreshold = try container.decode(Int.self, forKey: .highThreshold)
        stressThreshold = try container.decode(Int.self, forKey: .stressThreshold)

        if let value = try container.decodeIfPresent(Int.self, forKey: .repeats) {
            var items = [IDOWeekObjc]()
            for i in 0...6 {
                if ((value >> i) & 1) > 0 {
                    items.append(IDOWeekObjc(week: IDOWeek(rawValue: i)!))
                }
            }
            repeats = items
        } else {
            repeats = [IDOWeekObjc]()
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(onOff, forKey: .onOff)
        try container.encode(startHour, forKey: .startHour)
        try container.encode(startMinute, forKey: .startMinute)
        try container.encode(endHour, forKey: .endHour)
        try container.encode(endMinute, forKey: .endMinute)
        try container.encode(notifyFlag, forKey: .notifyFlag)
        try container.encode(remindOnOff, forKey: .remindOnOff)
        try container.encode(interval, forKey: .interval)
        try container.encode(highThreshold, forKey: .highThreshold)
        try container.encode(stressThreshold, forKey: .stressThreshold)

        var value = 0
        for item in repeats {
            value |= (1 << (item.week.rawValue + 1))
        }
        try container.encode(value, forKey: .repeats)
    }

    public func toJsonString() -> String? {
        if let jsonData = try? JSONEncoder().encode(self) {
            let jsonString = String(data: jsonData, encoding: .utf8)
            return jsonString
        }
        return nil
    }
}

File: ido_sdk_flutter/ido_sdk/protocol_sdk/protocol_channel/android/src/main/kotlin/com/idosmart/model/IDOStressSwitchParamModel.kt
package com.idosmart.model

import com.google.gson.GsonBuilder
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonSerializationContext
import com.google.gson.annotations.SerializedName
import java.lang.reflect.Type

open class IDOStressSwitchParamModel(
    onOff: Int,
    startHour: Int,
    startMinute: Int,
    endHour: Int,
    endMinute: Int,

    remindOnOff: Int,
    interval: Int,
    highThreshold: Int,
    stressThreshold: Int?,

    notifyFlag: Int,
    repeats: HashSet<IDOWeek>,
) : IDOBaseAdapterModel<IDOStressSwitchParamModel> {


    /// Overall switch 1: On 0: Off
    var onOff = onOff

    /// Start time - hour
    var startHour = startHour

    /// Start time - minute
    var startMinute = startMinute

    /// End time - hour
    var endHour = endHour

    /// End time - minute
    var endMinute = endMinute

    /// Stress reminder switch 1: On 0: Off
    /// Doesn't work if on_off is off
    var remindOnOff = remindOnOff

    /// Reminder interval in minutes, default is 60 minutes
    var interval = interval

    /// High pressure threshold
    var highThreshold = highThreshold

    /// Pressure calibration threshold, default is 80
    /// Requires firmware support for setSendCalibrationThreshold
    var stressThreshold = stressThreshold

    /** Notification type
     * 0: Invalid
     * 1: Allow notification
     * 2: Silent notification
     * 3: Disable notification
     * Requires firmware support for getPressureNotifyFlagMode
     */
    var notifyFlag = notifyFlag

    /// Repeat
    var repeats: HashSet<IDOWeek> = repeats


    override fun getDeSerializer(): IDOBaseModelDeSerializer<IDOStressSwitchParamModel>? {
        return StressSwitchParamModelSerializer()
    }

    override fun toString(): String {
        return GsonBuilder().create().toJson(this).toString()
    }
}


internal class StressSwitchParamModelSerializer : IDOBaseModelDeSerializer<IDOStressSwitchParamModel> {
    override fun serialize(src: IDOStressSwitchParamModel, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
        val jsonObject = JsonObject()
        jsonObject.addProperty("on_off", src.onOff)
        jsonObject.addProperty("start_hour", src.startHour)
        jsonObject.addProperty("start_minute", src.startMinute)
        jsonObject.addProperty("end_hour", src.endHour)
        jsonObject.addProperty("end_minute", src.endMinute)
        jsonObject.addProperty("interval", src.interval)
        jsonObject.addProperty("notify_flag", src.notifyFlag)
        jsonObject.addProperty("remind_on_off", src.remindOnOff)
        jsonObject.addProperty("high_threshold", src.highThreshold)
        jsonObject.addProperty("stress_threshold", src.stressThreshold)

        var num = 0
        src.repeats.forEach { week ->
            num = num or (1 shl week.rawValue + 1)
        }
        jsonObject.addProperty("repeat", num)
        return jsonObject
    }

    override fun deserialize(
        json: JsonElement?,
        typeOfT: Type?,
        context: JsonDeserializationContext?
    ): IDOStressSwitchParamModel? {
        json?.let {
            val jsonObject = it.asJsonObject
            val onOff = jsonObject.get("on_off").asInt
            val startHour = jsonObject.get("start_hour").asInt
            val startMinute = jsonObject.get("start_minute").asInt
            val endHour = jsonObject.get("end_hour").asInt
            val endMinute = jsonObject.get("end_minute").asInt
            val notifyFlag = jsonObject.get("notify_flag").asInt
            val interval = jsonObject.get("interval").asInt
            val remindOnOff = jsonObject.get("remind_on_off").asInt
            val highThreshold = jsonObject.get("high_threshold").asInt
            val stressThreshold = jsonObject.get("stress_threshold").asInt

            val repeat = jsonObject.get("repeat").asInt
            val items = HashSet<IDOWeek>()
            for (i in 0..6) {
                if (repeat and (1 shl (i + 1)) != 0) {
                    items.add(IDOWeek.fromRawValue(i))
                }
            }

            return IDOStressSwitchParamModel(onOff, startHour, startMinute, endHour, endMinute, remindOnOff, interval, highThreshold, stressThreshold, notifyFlag, items)
        }
        return null
    }
}


open class IDOStressSwitchModel(
    onOff: Int,
    startHour: Int,
    startMinute: Int,
    endHour: Int,
    endMinute: Int,

    remindOnOff: Int,
    interval: Int,
    highThreshold: Int,
    stressThreshold: Int?,

    notifyFlag: Int,
    repeats: HashSet<IDOWeek>,
) : IDOBaseAdapterModel<IDOStressSwitchModel> {


    /// Overall switch 1: On 0: Off
    var onOff = onOff

    /// Start time - hour
    var startHour = startHour

    /// Start time - minute
    var startMinute = startMinute

    /// End time - hour
    var endHour = endHour

    /// End time - minute
    var endMinute = endMinute

    /// Stress reminder switch 1: On 0: Off
    /// Doesn't work if on_off is off
    var remindOnOff = remindOnOff

    /// Reminder interval in minutes, default is 60 minutes
    var interval = interval

    /// High pressure threshold
    var highThreshold = highThreshold

    /// Pressure calibration threshold, default is 80
    /// Requires firmware support for setSendCalibrationThreshold
    var stressThreshold = stressThreshold

    /** Notification type
     * 0: Invalid
     * 1: Allow notification
     * 2: Silent notification
     * 3: Disable notification
     * Requires firmware support for getPressureNotifyFlagMode
     */
    var notifyFlag = notifyFlag

    /// Repeat
    var repeats: HashSet<IDOWeek> = repeats


    override fun getDeSerializer(): IDOBaseModelDeSerializer<IDOStressSwitchModel>? {
        return StressSwitchModelSerializer()
    }

    override fun toString(): String {
        return GsonBuilder().create().toJson(this).toString()
    }
}


internal class StressSwitchModelSerializer : IDOBaseModelDeSerializer<IDOStressSwitchModel> {
    override fun serialize(src: IDOStressSwitchModel, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
        val jsonObject = JsonObject()
        jsonObject.addProperty("on_off", src.onOff)
        jsonObject.addProperty("start_hour", src.startHour)
        jsonObject.addProperty("start_minute", src.startMinute)
        jsonObject.addProperty("end_hour", src.endHour)
        jsonObject.addProperty("end_minute", src.endMinute)
        jsonObject.addProperty("interval", src.interval)
        jsonObject.addProperty("notify_flag", src.notifyFlag)
        jsonObject.addProperty("remind_on_off", src.remindOnOff)
        jsonObject.addProperty("high_threshold", src.highThreshold)
        jsonObject.addProperty("stress_threshold", src.stressThreshold)

        var num = 0
        src.repeats.forEach { week ->
            num = num or (1 shl week.rawValue + 1)
        }
        jsonObject.addProperty("repeat", num)
        return jsonObject
    }

    override fun deserialize(
        json: JsonElement?,
        typeOfT: Type?,
        context: JsonDeserializationContext?
    ): IDOStressSwitchModel? {
        json?.let {
            val jsonObject = it.asJsonObject
            val onOff = jsonObject.get("on_off").asInt
            val startHour = jsonObject.get("start_hour").asInt
            val startMinute = jsonObject.get("start_minute").asInt
            val endHour = jsonObject.get("end_hour").asInt
            val endMinute = jsonObject.get("end_minute").asInt
            val notifyFlag = jsonObject.get("notify_flag").asInt
            val interval = jsonObject.get("interval").asInt
            val remindOnOff = jsonObject.get("remind_on_off").asInt
            val highThreshold = jsonObject.get("high_threshold").asInt
            val stressThreshold = jsonObject.get("stress_threshold").asInt

            val repeat = jsonObject.get("repeat").asInt
            val items = HashSet<IDOWeek>()
            for (i in 0..6) {
                if (repeat and (1 shl (i + 1)) != 0) {
                    items.add(IDOWeek.fromRawValue(i))
                }
            }

            return IDOStressSwitchModel(onOff, startHour, startMinute, endHour, endMinute, remindOnOff, interval, highThreshold, stressThreshold, notifyFlag, items)
        }
        return null
    }
}